lib/core/test/unit/Cardano/Pool/RankingSpec.hs

```diff
@@ -0,0 +1,254 @@
+
+            it "saturatedPoolRewards >= 0"
+                $ property prop_saturatedPoolRewardsNonNegative
+
+        describe "mkRelativeStake" $ do
+            it "stake > R, R >= 1 => mkRelativeStake ∈ [0,1] (even with conversions)"
+                $ property prop_mkRelativeStakeBounds
+
+prop_mkRelativeStakeBounds :: EpochConstants -> Word64 -> Word64 -> Property
+prop_mkRelativeStakeBounds constants' tot stake =
+    let
+        rel = unRelativeStake
+            $ mkRelativeStake
+                (Lovelace $ fromIntegral stake)
+                constants
+    in
+        counterexample ("relative stake = " ++ show rel) $
+            tot >= 1 ==>
+                tot >= stake ==>
+                    property $ rel >= 0 && rel <= 1
+
+  where
+    constants = constants'
+        { totalRewards = Lovelace . unsafeToNonNegative $ fromIntegral tot }
+
+prop_desirabilityNonNegative :: EpochConstants -> Pool -> Property
+prop_desirabilityNonNegative constants pool = property $
+    desirability constants pool >= 0
+
+prop_saturatedPoolRewardsNonNegative :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsNonNegative constants pool = property $
+    saturatedPoolRewards constants pool >= 0
+
+prop_saturatedPoolRewardsReduces :: EpochConstants -> Pool -> Property
+prop_saturatedPoolRewardsReduces constants' pool =
+    let
+        constants = constants'
+            { leaderStakeInfluence = unsafeToNonNegative 0 }
+        z0 = unRelativeStake  $ saturatedPoolSize constants
+        _R = R.getNonNegative $ getLovelace $ totalRewards constants
+        p  = R.getNonNegative $ recentAvgPerformance pool
+    in
+        saturatedPoolRewards constants pool === p*_R*z0
+
+prop_oversaturationDoesNotInfluenceDesirability
+    :: Pool
+    -> EpochConstants
+    -> Property
+prop_oversaturationDoesNotInfluenceDesirability pool constants =
+    forAll arbitrary $ \v1 v2 -> do
+        let (higherStake, lowerStake) = sortTuple (v1, v2)
+        let z0 = unRelativeStake $ saturatedPoolSize constants
+        higherStake > z0 ==>
+            des higherStake === des lowerStake
+  where
+    sortTuple (a, b)
+        | a > b     = (a, b)
+        | otherwise = (b, a)
+
+    des _stake = desirability constants pool
+    -- TODO: set pool{R.stake = stake}, but that field doesn't even exist!
+
+data EffectOnDesirability
+    = PositiveEffect
+    | NegativeEffect
+    deriving (Show, Eq)
+
+-- | All else equal, what effect on the desirability does changing a single
+-- field have?
+allElseEqualProperty
+    :: forall (field :: Symbol) a.
+        ( Arbitrary a
+        , Show a
+        , Ord a
+        , HasField' field Pool a
+        , KnownSymbol field
+        )
+    => EffectOnDesirability
+        -- ^ Expected effect on the desirability
+    -> Pool
+       -- ^ The base pool which is later modified
+    -> EpochConstants
+       -- ^ Constants used
+    -> Property
+allElseEqualProperty expectedEffect pool constants =
```

Comment:
Good one :+1:

Reply:
```Haskell
        , HasField' field Pool a
        , KnownSymbol field
```

Nice simplification 👍 , suspected that was theoretically possible